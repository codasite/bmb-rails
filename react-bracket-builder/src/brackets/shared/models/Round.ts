import { Nullable } from '../../../utils/types'
import { MatchNode } from './operations/MatchNode'
import { MatchRepr } from '../api/types/bracket'
import { Team } from './Team'

export class Round {
  index: number
  depth: number
  matches: Array<Nullable<MatchNode>>
  constructor(
    index: number,
    depth: number,
    matches: Array<Nullable<MatchNode>> = []
  ) {
    this.index = index
    this.depth = depth
    this.matches = matches
  }
  allPicked(): boolean {
    return this.matches.every((match) => {
      if (match === null) {
        return true
      }
      return match.team1Wins || match.team2Wins
    })
  }
  getTeam(teamId: number): Nullable<Team> {
    for (const [index, match] of this.matches.entries()) {
      if (match === null) {
        continue
      }
      const team1 = match.getTeam1()
      const team2 = match.getTeam2()

      if (team1?.id === teamId) {
        team1.side = index < this.matches.length / 2 ? 'left' : 'right'
        return team1
      }

      if (team2?.id === teamId) {
        team2.side = index < this.matches.length / 2 ? 'left' : 'right'
        return team2
      }
    }

    return null
  }
  serialize(): Nullable<MatchRepr>[] {
    return this.matches.map((match, i) => {
      if (match === null) {
        return null
      }
      return match.serialize()
    })
  }
}
