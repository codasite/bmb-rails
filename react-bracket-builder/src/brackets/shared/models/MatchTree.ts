import { Nullable } from '../../../utils/types'
import { BracketRes, MatchPick, MatchReq, MatchTreeRepr } from '../api'
import { WildcardPlacement } from './WildcardPlacement'
import { Team } from './Team'
import { MatchNode } from './operations/MatchNode'
import { Round } from './Round'
import { matchReprFromNumTeams } from './operations/MatchReprFromNumTeams'
import { getNumRounds } from './operations/GetNumRounds'
import { matchReprFromRes } from './operations/MatchReprFromRes'
import { linkNodes } from './operations/LinkNodes'

export class MatchTree {
  rounds: Round[]
  private numTeams: number
  private wildcardPlacement?: WildcardPlacement
  isVoting: boolean
  liveRoundIndex: number

  constructor(
    rounds: Round[] = [],
    wildcardPlacement?: WildcardPlacement,
    isVoting: boolean = false,
    liveRoundIndex: number = 0
  ) {
    linkNodes(rounds)
    this.rounds = rounds
    this.wildcardPlacement = wildcardPlacement
    this.isVoting = isVoting
    this.liveRoundIndex = liveRoundIndex
  }

  getRootMatch(): Nullable<MatchNode> {
    const lastRound = this.rounds[this.rounds.length - 1]
    if (!lastRound) {
      return null
    }
    return lastRound.matches[0]
  }

  clone(): MatchTree {
    return MatchTree.deserialize(this.serialize())
  }

  equals(other: MatchTree): boolean {
    return (
      JSON.stringify(this.serialize()) === JSON.stringify(other.serialize())
    )
  }

  /**
   * Returns the total number of POSSIBLE teams in the tournament.
   * This should reflect the number of teams that would be used to create the same structure with MatchTree.fromNumTeams()
   */
  getNumTeams(): number {
    if (this.numTeams) {
      return this.numTeams
    }
    // count all non null matches in the tree
    const numMatches = this.rounds.reduce((acc, round) => {
      return (
        acc +
        round.matches.reduce((acc, match) => {
          if (match) {
            return acc + 1
          }
          return acc
        }, 0)
      )
    }, 0)

    const numTeams = numMatches + 1
    this.numTeams = numTeams

    return numTeams
  }

  getWildcardPlacement(): WildcardPlacement | undefined {
    return this.wildcardPlacement
  }

  serialize(): MatchTreeRepr {
    const tree = this
    const rounds = tree.rounds.map((round) => {
      return round.serialize()
    })
    return {
      rounds: rounds,
      wildcardPlacement: tree.wildcardPlacement,
      isVoting: tree.isVoting,
      liveRoundIndex: tree.liveRoundIndex,
    }
  }

  advanceTeam = (
    roundIndex: number,
    matchIndex: number,
    left: boolean,
    requireComplete: boolean = false
  ) => {
    if (requireComplete && roundIndex > 0) {
      const prevRound = this.rounds[roundIndex - 1]
      if (prevRound && !prevRound.allPicked()) {
        return
      }
    }
    const round = this.rounds[roundIndex]
    const match = round.matches[matchIndex]
    if (!match) {
      return
    }
    match.team1Wins = false
    match.team2Wins = false
    if (left) {
      match.team1Wins = true
    } else {
      match.team2Wins = true
    }
  }

  anyPicked = (): boolean => {
    return this.rounds.some((round) => {
      return round.matches.some((match) => {
        if (!match) {
          return false
        }
        return match.team1Wins || match.team2Wins
      })
    })
  }

  allPicked = (): boolean => {
    if (this.isVoting) {
      return this.rounds[this.liveRoundIndex].allPicked()
    }
    return this.rounds.every((round) => {
      return round.allPicked()
    })
  }

  allTeamsAdded = (): boolean => {
    return this.everyMatch((match) => {
      let hasNeededTeams = true
      if (!match.left && !match.getTeam1()?.name) {
        hasNeededTeams = false
      }
      if (hasNeededTeams && !match.right && !match.getTeam2()?.name) {
        hasNeededTeams = false
      }
      return hasNeededTeams
    })
  }

  toMatchReq = (): MatchReq[] => {
    const matches: MatchReq[] = []
    this.forEachMatch((match, matchIndex, roundIndex) => {
      const matchReq: MatchReq = {
        roundIndex: match.roundIndex,
        matchIndex: match.matchIndex,
      }
      if (!match.left) {
        matchReq.team1 = match.getTeam1()?.toTeamReq()
      }
      if (!match.right) {
        matchReq.team2 = match.getTeam2()?.toTeamReq()
      }
      if (matchReq.team1 || matchReq.team2) {
        matches.push(matchReq)
      }
    })
    return matches
  }

  toMatchPicks = (): MatchPick[] => {
    const picks: MatchPick[] = []
    this.rounds.forEach((round, roundIndex) => {
      round.matches.forEach((match, matchIndex) => {
        if (!match) {
          return
        }
        const { team1Wins, team2Wins } = match
        const team1 = match.getTeam1()
        const team2 = match.getTeam2()
        const team1Winner = team1Wins && team1
        const team2Winner = team2Wins && team2

        if (team1Winner && team1.id) {
          picks.push({
            roundIndex,
            matchIndex,
            winningTeamId: team1.id,
          })
        } else if (team2Winner && team2.id) {
          picks.push({
            roundIndex,
            matchIndex,
            winningTeamId: team2.id,
          })
        }
      })
    })
    return picks
  }

  forEachMatch = (
    callback: (match: MatchNode, matchIndex: number, roundIndex: number) => void
  ) => {
    this.rounds.forEach((round, roundIndex) => {
      round.matches.forEach((match, matchIndex) => {
        if (!match) {
          return
        }
        callback(match, matchIndex, roundIndex)
      })
    })
  }

  everyMatch = (
    callback: (
      match: MatchNode,
      matchIndex: number,
      roundIndex: number
    ) => boolean
  ) => {
    return this.rounds.every((round, roundIndex) => {
      return round.matches.every((match, matchIndex) => {
        if (!match) {
          return true
        }
        return callback(match, matchIndex, roundIndex)
      })
    })
  }

  findMatch = (
    callback: (match: Nullable<MatchNode>) => boolean
  ): Nullable<MatchNode> => {
    let foundMatch: Nullable<MatchNode> = null
    this.rounds.some((round) => {
      return round.matches.some((match) => {
        if (callback(match)) {
          foundMatch = match
          return true
        }
        return false
      })
    })
    return foundMatch
  }

  static fromNumTeams(
    numTeams: number,
    wildcardPlacement: WildcardPlacement = WildcardPlacement.Top
  ): MatchTree {
    const matches = matchReprFromNumTeams(numTeams, wildcardPlacement)
    return MatchTree.deserialize({
      rounds: matches,
      wildcardPlacement,
    })
  }

  static fromMatchRes(
    bracket: Pick<
      BracketRes,
      'matches' | 'numTeams' | 'isVoting' | 'liveRoundIndex'
    >,
    wildcardPlacement?: WildcardPlacement
  ): MatchTree | null {
    const numRounds = getNumRounds(bracket.numTeams)

    const nestedMatches = matchReprFromRes(numRounds, bracket.matches)
    return MatchTree.deserialize({
      rounds: nestedMatches,
      wildcardPlacement,
      isVoting: bracket.isVoting,
      liveRoundIndex: bracket.liveRoundIndex,
    })
  }

  static fromPicks(
    bracket: Pick<
      BracketRes,
      'matches' | 'numTeams' | 'isVoting' | 'liveRoundIndex'
    >,
    picks: MatchPick[],
    wildcardPlacement?: WildcardPlacement
  ): MatchTree | null {
    const matchTree = MatchTree.fromMatchRes(bracket, wildcardPlacement)
    if (!matchTree) {
      return null
    }
    for (const pick of picks) {
      const { roundIndex, matchIndex, winningTeamId } = pick
      const match = matchTree.rounds[roundIndex].matches[matchIndex]
      if (!match) {
        throw new Error('Match not found')
      }
      match._pick = pick

      const team1 = match.getTeam1()
      const team2 = match.getTeam2()
      if (team1?.id === winningTeamId) {
        match.team1Wins = true
      } else if (team2?.id === winningTeamId) {
        match.team2Wins = true
      } else {
        throw new Error('Match does not have winningTeam from pick')
      }
    }
    return matchTree
  }

  static deserialize(matchTreeRepr: MatchTreeRepr): MatchTree {
    const { rounds: matchRes, wildcardPlacement } = matchTreeRepr
    const rounds = matchRes.map((round, roundIndex) => {
      const depth = matchRes.length - roundIndex - 1
      const newRound = new Round(roundIndex, depth)
      const matches = round.map((match, matchIndex) => {
        if (match === null) {
          return null
        }
        const {
          id,
          team1: team1Repr,
          team2: team2Repr,
          team1Wins,
          team2Wins,
          pick,
        } = match
        const team1 = team1Repr ? new Team(team1Repr.name, team1Repr.id) : null
        const team2 = team2Repr ? new Team(team2Repr.name, team2Repr.id) : null
        // const newMatch = new MatchNode(roundIndex, matchIndex, depth, match.id, team1, team2, team1Wins, team2Wins)
        const newMatch = new MatchNode({
          id,
          matchIndex,
          roundIndex,
          team1,
          team2,
          team1Wins,
          team2Wins,
          left: null,
          right: null,
          parent: null,
          depth,
          pick,
        })
        return newMatch
      })
      newRound.matches = matches
      return newRound
    })
    const tree = new MatchTree(
      rounds,
      wildcardPlacement,
      matchTreeRepr.isVoting,
      matchTreeRepr.liveRoundIndex
    )
    return tree
  }
}
