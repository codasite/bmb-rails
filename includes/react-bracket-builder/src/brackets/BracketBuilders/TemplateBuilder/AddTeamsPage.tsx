import React, { } from 'react'
import { MatchTree } from '../../shared/models/MatchTree'
import { ReactComponent as ArrowNarrowLeft } from '../../shared/assets/arrow-narrow-left.svg'
import iconBackground from '../../shared/assets/bmb_icon_white_02.png'
import { BracketMeta, BracketMetaContext, DarkModeContext } from '../../shared/context'
import { AddTeamsBracket } from '../../shared/components/Bracket'
import { ActionButton } from '../../shared/components/ActionButtons'
import { ReactComponent as ShuffleIcon } from '../../shared/assets/shuffle.svg'
import { ReactComponent as SaveIcon } from '../../shared/assets/save.svg'
import { ReactComponent as PlayIcon } from '../../shared/assets/play.svg'

interface AddTeamsPageProps {
	matchTree?: MatchTree
	setMatchTree?: (matchTree: MatchTree) => void
	handleSaveTemplate: () => void
	handleSaveTournament: () => void
	handleBack: () => void
}

export const AddTeamsPage = (props: AddTeamsPageProps) => {
	const {
		matchTree,
		setMatchTree,
		handleSaveTemplate,
		handleSaveTournament,
		handleBack
	} = props

	const createDisabled = !matchTree || !matchTree.allTeamsAdded()

	return (
		<div className='tw-flex tw-flex-col tw-gap-60 tw-pt-30 tw-pb-60 tw-bg-no-repeat tw-bg-top tw-bg-cover' style={{ 'background': `url(${iconBackground}), #000225` }}>
			<div className='tw-px-60'>
				<div className='tw-flex tw-p-16' >
					<a href="#" className='tw-flex tw-gap-10 tw-items-center' onClick={handleBack}>
						<ArrowNarrowLeft />
						<span className='tw-font-500 tw-text-20 tw-text-white '>Create Template</span>
					</a>
				</div>
			</div>
			<div className={`tw-flex tw-flex-col tw-justify-center tw-items-center tw-max-w-screen-xl tw-min-h-[500px] tw-m-auto tw-dark`}>
				{
					matchTree &&
					<AddTeamsBracket
						matchTree={matchTree}
						setMatchTree={setMatchTree}
					/>
				}
			</div>
			<div className='tw-flex tw-flex-col tw-gap-[46px] tw-max-w-screen-lg tw-m-auto tw-w-full'>
				<ActionButton className='tw-self-center' variant='blue' onClick={handleBack} paddingX={16} paddingY={12}>
					<ShuffleIcon />
					<span className='tw-font-500 tw-text-20 tw-uppercase tw-font-sans'>Scramble Team Order</span>
				</ActionButton>
				<div className='tw-flex tw-flex-col tw-gap-16'>
					<ActionButton variant='blue' gap={16} disabled={createDisabled} onClick={handleSaveTemplate}>
						<SaveIcon />
						<span className='tw-font-500 tw-text-20 tw-uppercase tw-font-sans'>Save As Template</span>
					</ActionButton>
					{/* <ActionButton variant='green' gap={16} disabled={createDisabled} onClick={handleSaveTournament}> */}
					<ActionButton variant='green' gap={16} disabled={true} onClick={handleSaveTournament}>
						<PlayIcon />
						<span className='tw-font-500 tw-text-20 tw-uppercase tw-font-sans'>Create Tournament</span>
					</ActionButton>
				</div>

			</div>
		</div>
	)
}