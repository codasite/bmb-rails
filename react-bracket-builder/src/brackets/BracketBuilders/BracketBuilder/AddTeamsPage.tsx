// DO NOT REMOVE REACT IMPORT. Needed for image generator
import React, { useContext, useState } from 'react'
import { MatchTree } from '../../shared/models/MatchTree'
import { ReactComponent as ArrowNarrowLeft } from '../../shared/assets/arrow-narrow-left.svg'
import { AddTeamsBracket } from '../../shared/components/Bracket'
import { ActionButton } from '../../shared/components/ActionButtons'
import { ReactComponent as SaveIcon } from '../../shared/assets/save.svg'
import { getBracketWidth } from '../../shared/components/Bracket/utils'
import { DatePicker } from '../../shared/components/DatePicker'
import { WindowDimensionsContext } from '../../shared/context/WindowDimensionsContext'
import { WithWindowDimensions } from '../../shared/components/HigherOrder/WithWindowDimensions'
import { BracketBuilderHeader } from './BracketBuilderHeader'
import { AddTeamsFullBracketPage } from './PaginatedAddTeams/AddTeamsFullBracketPage'
import { AddTeamsPages } from './PaginatedAddTeams/AddTeamsPages'
import { PaginatedBuilder } from '../PaginatedBuilderBase/PaginatedBuilder'
import { BracketBackground } from '../../shared/components/BracketBackground'
import { ScrambleButton } from './ScrambleButton'
import { BracketMetaContext } from '../../shared/context/context'
interface AddTeamsPageProps {
  matchTree?: MatchTree
  setMatchTree?: (matchTree: MatchTree) => void
  handleSaveBracket: () => void
  handleBack: () => void
  month?: string
  setMonth?: (month: string) => void
  year?: string
  setYear?: (year: string) => void
  processing?: boolean
}
const AddTeamsPage = (props: AddTeamsPageProps) => {
  const {
    matchTree,
    setMatchTree,
    handleSaveBracket,
    handleBack,
    month,
    setMonth,
    year,
    setYear,
    processing,
  } = props
  const [dateError, setDateError] = useState(false)
  const { bracketMeta } = useContext(BracketMetaContext)
  const createDisabled =
    !matchTree ||
    !matchTree.allTeamsAdded() ||
    dateError ||
    processing ||
    bracketMeta?.title.length === 0
  const { width: windowWidth } = useContext(WindowDimensionsContext)
  const showPaginated = windowWidth < getBracketWidth(matchTree.rounds.length)
  return (
    <BracketBackground>
      <BracketBuilderHeader />
      <div className="sm:tw-px-60 tw-mb-16">
        <div className="tw-flex tw-p-16">
          <a
            href="#"
            className="tw-flex tw-gap-10 tw-items-center"
            onClick={handleBack}
          >
            <ArrowNarrowLeft />
            <span className="tw-font-500 tw-text-16 sm:tw-text-20 tw-text-white ">
              Create Bracket
            </span>
          </a>
        </div>
      </div>
      <div
        className={`tw-flex tw-flex-col tw-w-full tw-max-w-screen-lg tw-mx-auto ${
          showPaginated ? 'tw-gap-30' : 'tw-gap-60'
        }`}
      >
        <div>
          <div
            className={`tw-flex tw-flex-col tw-justify-center tw-items-center tw-dark ${
              showPaginated ? 'tw-overflow-hidden' : ''
            }`}
          >
            {matchTree && !showPaginated && (
              <AddTeamsBracket
                matchTree={matchTree}
                setMatchTree={setMatchTree}
              />
            )}
            {matchTree && showPaginated && (
              <PaginatedBuilder
                matchTree={matchTree}
                setMatchTree={setMatchTree}
                BracketPagesComponent={AddTeamsPages}
                EndPageComponent={AddTeamsFullBracketPage}
                initialPage={createDisabled ? 'bracket' : 'end'}
              />
            )}
          </div>
        </div>
        {matchTree && setMatchTree && matchTree.rounds.length < 7 && (
          <ScrambleButton
            matchTree={matchTree}
            setMatchTree={setMatchTree}
            showPaginated={showPaginated}
            processing={processing}
          />
        )}
        <div className="tw-flex tw-flex-col tw-gap-60 tw-max-w-[510px] tw-w-full tw-mx-auto">
          <DatePicker
            month={month}
            year={year}
            handleMonthChange={(month) => setMonth(month)}
            handleYearChange={(year) => setYear(year)}
            onHasError={(error) => setDateError(true)}
            onErrorCleared={() => setDateError(false)}
            showTitle={true}
          />
        </div>
        <ActionButton
          variant="green"
          gap={16}
          disabled={createDisabled}
          onClick={handleSaveBracket}
        >
          <SaveIcon />
          <span className="tw-font-500 tw-text-20 tw-uppercase tw-font-sans">
            Save
          </span>
        </ActionButton>
      </div>
    </BracketBackground>
  )
}

const Wrapped = WithWindowDimensions(AddTeamsPage)
export { Wrapped as AddTeamsPage }
