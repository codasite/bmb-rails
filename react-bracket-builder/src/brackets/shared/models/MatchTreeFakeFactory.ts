import { MatchTree } from './MatchTree'

export function fourTeamMatchTree() {
  const matches = [
    [
      {
        id: 9,
        roundIndex: 0,
        matchIndex: 0,
        team1: { id: 17, name: 'Team 1' },
        team2: { id: 18, name: 'Team 2' },
        team1Wins: true,
      },
      {
        id: 10,
        roundIndex: 0,
        matchIndex: 1,
        team1: { id: 19, name: 'Team 3' },
        team2: { id: 20, name: 'Team 4' },
        team1Wins: true,
      },
    ],
    [{ roundIndex: 1, matchIndex: 1, team2Wins: true }],
  ]
  return MatchTree.deserialize({ rounds: matches })
}
