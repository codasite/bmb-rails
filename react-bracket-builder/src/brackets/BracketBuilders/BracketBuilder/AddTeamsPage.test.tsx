import { render } from '@testing-library/react'
import { AddTeamsPage } from './AddTeamsPage'
import { fourTeamMatchTree } from '../../shared/models/MatchTreeFakeFactory'

describe('AddTeamsPage', () => {
  test('renders page correctly', () => {
    const { asFragment } = render(
      <AddTeamsPage
        handleBack={() => {}}
        handleSaveBracket={() => {}}
        matchTree={fourTeamMatchTree()}
      />
    )
    expect(asFragment()).toMatchSnapshot()
  })
})
