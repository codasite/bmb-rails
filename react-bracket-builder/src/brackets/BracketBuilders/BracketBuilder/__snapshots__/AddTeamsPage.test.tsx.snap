// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AddTeamsPage renders page correctly 1`] = `
<DocumentFragment>
  <div
    class="tw-flex tw-flex-col tw-pt-30 tw-pb-60 tw-bg-no-repeat tw-bg-top tw-bg-cover tw-px-16 sm:tw-px-20"
  >
    <div
      class="tw-py-30"
    >
      <div
        class="tw-flex tw-flex-col tw-items-center tw-gap-20"
      >
        <div
          class="tw-h-[50px] tw-w-[200px] md:tw-h-[75px] md:tw-w-[300px] lg:tw-h-[85px] lg:tw-w-[340px] tw-bg-[url('https://s3.amazonaws.com/backmybracket.com/bmb_text_logo_500.png')] tw-bg-no-repeat tw-bg-contain"
        />
        <h1
          class="tw-text-32 sm:tw-text-48 md:tw-text-64 lg:tw-text-80 tw-text-center"
        >
          Bracket Builder
        </h1>
      </div>
    </div>
    <div
      class="sm:tw-px-60 tw-mb-16"
    >
      <div
        class="tw-flex tw-p-16"
      >
        <a
          class="tw-flex tw-gap-10 tw-items-center"
          href="#"
        >
          <svg />
          <span
            class="tw-font-500 tw-text-16 sm:tw-text-20 tw-text-white "
          >
            Create Bracket
          </span>
        </a>
      </div>
    </div>
    <div
      class="tw-flex tw-flex-col tw-w-full tw-max-w-screen-lg tw-mx-auto tw-gap-60"
    >
      <div>
        <div
          class="tw-flex tw-flex-col tw-justify-center tw-items-center tw-dark "
        >
          <div
            class="tw-flex tw-flex-col tw-dark wpbb-default-bracket tw-relative"
          >
            <div
              class="tw-flex tw-flex-col tw-justify-end"
              style="margin-bottom: 95px; min-height: 0;"
            >
              <div
                class="tw-flex tw-flex-col tw-items-center"
                style="gap: 24px;"
              >
                <div
                  class="wpbb-team-1-0-winner tw-relative tw-flex tw-justify-center tw-items-center tw-whitespace-nowrap tw-leading-none tw-uppercase tw-font-sans tw-w-[297px] tw-h-[52px] tw-text-white tw-font-700 tw-bg-transparent tw-border-solid tw-border-white/25"
                  data-testid="team-slot-round-1-match-0-winner"
                  style="border-width: 2px;"
                >
                  <span
                    style="font-size: 32px; transform: scale(1); visibility: hidden;"
                  >
                    Team 3
                  </span>
                </div>
              </div>
            </div>
            <div
              class="tw-flex tw-flex-col tw-justify-center tw-h-100"
            >
              <div
                class="tw-flex tw-justify-between"
                style="width: 700px;"
              >
                <div
                  class="wpbb-default-match-column tw-flex tw-flex-col tw-justify-center tw-gap-[80px]"
                >
                  <div
                    class="tw-flex tw-flex-col tw-gap-[20px] tw-translate-y-[0px]"
                  >
                    <div
                      class="wpbb-team-0-0-left tw-relative tw-flex tw-justify-center tw-items-center tw-whitespace-nowrap tw-leading-none tw-uppercase tw-font-sans tw-w-[115px] tw-h-[28px] tw-text-white tw-font-500 tw-bg-transparent tw-border-solid tw-border-white/25"
                      data-testid="team-slot-round-0-match-0-left"
                      style="border-width: 2px;"
                    >
                      <span
                        style="font-size: 16px; transform: scale(1); visibility: hidden;"
                      >
                        Team 1
                      </span>
                    </div>
                    <div
                      class="wpbb-team-0-0-right tw-relative tw-flex tw-justify-center tw-items-center tw-whitespace-nowrap tw-leading-none tw-uppercase tw-font-sans tw-w-[115px] tw-h-[28px] tw-text-white tw-font-500 tw-bg-transparent tw-border-solid tw-border-white/25"
                      data-testid="team-slot-round-0-match-0-right"
                      style="border-width: 2px;"
                    >
                      <span
                        style="font-size: 16px; transform: scale(1); visibility: hidden;"
                      >
                        Team 2
                      </span>
                    </div>
                  </div>
                </div>
                <div
                  class="wpbb-default-match-column tw-flex tw-flex-col tw-justify-center tw-gap-[236px]"
                >
                  <div
                    class="tw-flex tw-flex-col tw-gap-[20px] tw-translate-y-[-48px]"
                  >
                    <div
                      class="wpbb-team-1-0-left tw-relative tw-flex tw-justify-center tw-items-center tw-whitespace-nowrap tw-leading-none tw-uppercase tw-font-sans tw-w-[115px] tw-h-[28px] tw-text-white tw-font-500 tw-bg-transparent tw-border-solid tw-border-white/25"
                      data-testid="team-slot-round-1-match-0-left"
                      style="border-width: 2px;"
                    >
                      <span
                        style="font-size: 16px; transform: scale(1); visibility: hidden;"
                      >
                        Team 1
                      </span>
                    </div>
                    <div
                      class="wpbb-team-1-0-right tw-relative tw-flex tw-justify-center tw-items-center tw-whitespace-nowrap tw-leading-none tw-uppercase tw-font-sans tw-w-[115px] tw-h-[28px] tw-text-white tw-font-500 tw-bg-transparent tw-border-solid tw-border-white/25"
                      data-testid="team-slot-round-1-match-0-right"
                      style="border-width: 2px;"
                    >
                      <span
                        style="font-size: 16px; transform: scale(1); visibility: hidden;"
                      >
                        Team 3
                      </span>
                    </div>
                  </div>
                </div>
                <div
                  class="wpbb-default-match-column tw-flex tw-flex-col tw-justify-center tw-gap-[80px]"
                >
                  <div
                    class="tw-flex tw-flex-col tw-gap-[20px] tw-translate-y-[0px]"
                  >
                    <div
                      class="wpbb-team-0-1-left tw-relative tw-flex tw-justify-center tw-items-center tw-whitespace-nowrap tw-leading-none tw-uppercase tw-font-sans tw-w-[115px] tw-h-[28px] tw-text-white tw-font-500 tw-bg-transparent tw-border-solid tw-border-white/25"
                      data-testid="team-slot-round-0-match-1-left"
                      style="border-width: 2px;"
                    >
                      <span
                        style="font-size: 16px; transform: scale(1); visibility: hidden;"
                      >
                        Team 3
                      </span>
                    </div>
                    <div
                      class="wpbb-team-0-1-right tw-relative tw-flex tw-justify-center tw-items-center tw-whitespace-nowrap tw-leading-none tw-uppercase tw-font-sans tw-w-[115px] tw-h-[28px] tw-text-white tw-font-500 tw-bg-transparent tw-border-solid tw-border-white/25"
                      data-testid="team-slot-round-0-match-1-right"
                      style="border-width: 2px;"
                    >
                      <span
                        style="font-size: 16px; transform: scale(1); visibility: hidden;"
                      >
                        Team 4
                      </span>
                    </div>
                  </div>
                </div>
              </div>
            </div>
            <div
              style="margin-top: 20px; min-height: 0;"
            >
              <div
                class="tw-flex tw-flex-col tw-gap-20 tw-justify-between tw-items-center tw-font-700 tw-whitespace-nowrap "
              >
                <span
                  class="tw-text-36 tw-text-dd-blue dark:tw-text-white"
                >
                  Who You Got?
                </span>
                <svg
                  class="tw-w-[124px] tw-text-black/25 dark:tw-text-white"
                />
              </div>
            </div>
            <div
              class="tw-absolute"
            />
            <div
              class="tw-absolute"
            />
          </div>
        </div>
      </div>
      <div
        class="tw-flex tw-flex-col tw-justify-center tw-gap-10"
      >
        <button
          class="tw-flex tw-flex-row tw-items-center tw-justify-center tw-font-sans tw-uppercase tw-whitespace-nowrap tw-cursor-pointer disabled:tw-cursor-default tw-leading-none tw-cursor-pointer tw-bg-blue/15 tw-text-white tw-gap-10 tw-text-20 tw-font-500 tw-rounded-8 tw-px-16 tw-py-12 tw-border-solid tw-border tw-border-blue tw-self-center"
        >
          <svg />
          <span
            class="tw-font-500 tw-text-20 tw-uppercase tw-font-sans"
          >
            Scramble Team Order
          </span>
        </button>
      </div>
      <div
        class="tw-flex tw-flex-col tw-gap-60 tw-max-w-[510px] tw-w-full tw-mx-auto"
      >
        <div
          class="tw-flex tw-flex-col tw-justify-center tw-text-center tw-gap-16"
        >
          <span
            class="tw-text-white/50 tw-text-24 tw-font-500"
          >
            Your Bracket's Date
          </span>
          <div
            class="tw-flex tw-flex-col sm:tw-items-start sm:tw-flex-row tw-justify-center tw-gap-16 "
          >
            <div
              class="tw-flex tw-flex-col"
            >
              <div
                class="tw-relative tw-flex tw-flex-col tw-gap-8"
              >
                <div
                  class="tw-absolute tw-top-1/2 tw-left-1/2 tw--translate-x-1/2 tw--translate-y-1/2 tw-pointer-events-none"
                >
                  <div
                    class="tw-flex tw-items-center tw-justify-center tw-gap-16 tw-pointer-events-none"
                  >
                    <svg />
                    <span
                      class="tw-text-24 tw-font-600 tw-text-white/50"
                    >
                      Month
                    </span>
                  </div>
                </div>
                <input
                  class="tw-font-sans tw-uppercase tw-p-16 tw-border tw-border-solid tw-rounded-8 tw-border-white/50 tw-text-white/50 tw-text-center tw-text-24 tw-font-600 tw-text-white-50 tw-placeholder-white/50 tw-bg-transparent focus:tw-placeholder-transparent focus:tw-outline-none focus:tw-bg-white/5 tw-flex-grow "
                  type="text"
                  value=""
                />
              </div>
            </div>
            <div
              class="tw-relative tw-flex tw-flex-col tw-gap-8"
            >
              <input
                class="tw-font-sans tw-uppercase tw-p-16 tw-border tw-border-solid tw-rounded-8 tw-border-white/50 tw-text-white/50 tw-text-center tw-text-24 tw-font-600 tw-text-white-50 tw-placeholder-white/50 tw-bg-transparent focus:tw-placeholder-transparent focus:tw-outline-none focus:tw-bg-white/5 sm:tw-w-[150px] "
                type="text"
                value=""
              />
            </div>
          </div>
        </div>
      </div>
      <button
        class="tw-flex tw-flex-row tw-items-center tw-justify-center tw-font-sans tw-uppercase tw-whitespace-nowrap tw-cursor-pointer disabled:tw-cursor-default tw-leading-none tw-cursor-pointer tw-bg-blue/15 tw-text-white tw-gap-16 tw-text-20 tw-font-500 tw-rounded-8 tw-py-16 tw-border-solid tw-border tw-border-blue"
      >
        <svg />
        <span
          class="tw-font-500 tw-text-20 tw-uppercase tw-font-sans"
        >
          Save
        </span>
      </button>
    </div>
  </div>
</DocumentFragment>
`;
